File: src/main/java/com/github/lipinskipawel/BruteForceThinking.java
package com.github.lipinskipawel;

import com.github.lipinskipawel.board.ai.MoveStrategy;
import com.github.lipinskipawel.board.engine.BoardInterface;
import com.github.lipinskipawel.board.engine.Move;
import com.github.lipinskipawel.gui.GameDrawer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.swing.*;
import java.util.concurrent.atomic.AtomicBoolean;

public final class BruteForceThinking extends SwingWorker<Move, Integer> {

    private final MoveStrategy moveStrategy;
    private final BoardInterface board;
    private final int depth;
    private final GameDrawer gameDrawer;
    private AtomicBoolean canHumanMove;
    private final Logger logger = LoggerFactory.getLogger(BruteForceThinking.class);

    public BruteForceThinking(final MoveStrategy moveStrategy,
                              final BoardInterface board,
                              final int depth,
                              final GameDrawer gameDrawer,
                              final AtomicBoolean canHumanMove) {
        this.moveStrategy = moveStrategy;
        this.board = board;
        this.depth = depth;
        this.gameDrawer = gameDrawer;
        this.canHumanMove = canHumanMove;
    }

    @Override
    protected Move doInBackground() {
        logger.info("searching for move");
        final var bestMove = moveStrategy.execute(this.board, this.depth);
        this.gameDrawer.drawMove(board.executeMove(bestMove), board.getPlayer().opposite());
        this.canHumanMove.set(true);
        logger.info("set canHumanMove to : " + this.canHumanMove);
        return bestMove;
    }


}

File: football-gui-game/src/main/kotlin/com/github/lipinskipawel/BruteForceThinking.kt
package com.github.lipinskipawel

import com.github.lipinskipawel.board.ai.MoveStrategy
import com.github.lipinskipawel.board.engine.BoardInterface
import com.github.lipinskipawel.board.engine.Move
import com.github.lipinskipawel.gui.DrawableFootballPitch
import org.slf4j.LoggerFactory
import java.util.concurrent.atomic.AtomicBoolean
import javax.swing.SwingWorker

class BruteForceThinking(
        private val moveStrategy: MoveStrategy,
        private val board: BoardInterface,
        private val depth: Int,
        private val gameDrawer: DrawableFootballPitch,
        private val canHumanMove: AtomicBoolean) : SwingWorker<Move, Int>() {

    companion object {
        private val logger = LoggerFactory.getLogger(BruteForceThinking::class.java)
    }

    override fun doInBackground(): Move {
        logger.info("searching for move")
        val bestMove = moveStrategy.execute(board, depth)
        if (Thread.currentThread().isInterrupted) {
            logger.info("ai has been interrupted. No more work to be done")
        } else {
            gameDrawer.drawPitch(board.executeMove(bestMove), board.player.opposite())
            logger.info("set canHumanMove to : $canHumanMove")
        }
        canHumanMove.set(true)
        return bestMove
    }
}